algebra2/gaussian_elimination/gaussian_elimination.hpp

```diff
@@ -0,0 +1,500 @@
+    }
+
+    enum class axis {
+        rows,
+        columns,
+    };
+
+    enum class reducted_form {
+        echelon,
+        diagonal
+    };
+
+    enum class error {
+        not_invertible,
+        not_square,
+    };
+    
+    /*
+        description:
+            performs gaussian elimination on matrix m and returns it's determinant, reduced matrix and steps
+    */
+    template <typename T, typename LP>
+    auto gaussian_elimiantion_alg(ranges::matrix_view<T, LP> m, 
+                                    axis a, 
+                                    reducted_form reducted,
+                                    operation allowed_operations = operation::swap | operation::add | operation::multiply){
+        auto [fracs, matrix_of_fracs] = convert_to_matrix_of_fractions(m);
+
+        int rows = matrix_of_fracs.extent(0),
+                    cols = matrix_of_fracs.extent(1);
+        fraction<int> factor;
+        std::pair<std::vector<std::string>, fraction<int>> values;
+        values.second = {1, 1};      
+
+        for(int i = 0;  i < rows; i++){
+            if(matrix_of_fracs[i, i].numerator == 0){
+                for(int j = i+1; j < rows; j++){
+                    if(matrix_of_fracs[j, i].numerator != 0){
+                        swap(m, i, j);
+                        values.first.push_back(std::format("swap R{} with R{}", j+1, i+1));
+                        values.second *= -1; 
+                    }
+                }
+            }
+            if(matrix_of_fracs[i, i].numerator != 0){
+                for(int k = i+1; k < rows; k++){
+                    factor = matrix_of_fracs[k, i]/matrix_of_fracs[i, i];
+                    if(factor.numerator != 0){
+                        subtract(matrix_of_fracs, k, i, factor);
+                        values.first.push_back(std::format("R{} - {} * R{}", k+1, factor, i+1)); 
+                    }
+                }
+            }
+        }
+        if(reducted == reducted_form::diagonal){
+            for(int i = 0;  i < rows; i++){
+                if(matrix_of_fracs[i, i].numerator != 0){
+                    for(int k = i+1; k < rows; k++){
+                        factor = matrix_of_fracs[i, k]/matrix_of_fracs[k, k];
+                        subtract(matrix_of_fracs, i, k, factor);
+                    }
+                }
+            }
+
+            for(int i = 0; i < rows; i++){
+                factor = matrix_of_fracs[i, i];
+                matrix_of_fracs[i] /= factor;
+                values.second *= factor * -1; 
+            }
+        }
+        return std::pair{std::pair{std::move(fracs), matrix_of_fracs}, values};
+    }
+
+    /*
+        description:
+            performs gaussian elimination on matrix m and returns it's determinant, reduced matrix and steps(overloader for matrix with fractions)
+    */
+    template <std::integral T, typename LP>
+    auto gaussian_elimiantion_alg(ranges::matrix_view<fraction<T>, LP> m, 
```

Comment:
code duplication - try to not provide specializations for fraction - code looks very similar - adjust lines which differ